corehq/apps/hqwebapp/views.py

```diff
@@ -1543,3 +1545,27 @@ def check_sso_login_status(request):
         'sso_url': sso_url,
         'continue_text': continue_text,
     })
+
+
+@require_POST
+def set_language(request):
+    """
+    Redirect to the current page while setting the chosen language, if valid.
+    If no http referer is available, just set the language. Based on
+    django.views.i18n.set_language.
+    """
+    next_url = request.META.get("HTTP_REFERER")
+    response = HttpResponseRedirect(next_url) if next_url else HttpResponse(status=204)
+    lang_code = request.POST.get("language")
+    if lang_code and lang_code in langs_by_code.keys():
```

Comment:
One thing I'm not sure of is if this should validate against `langcodes` -- that is, if it's anything we consider a valid langcode even if we don't have translations for it -- or if it should validate against `settings.LANGUAGES`, which specifies languages we have translations for.

A user will only see the options in the dropdown that are defined in `settings.LANGUAGES` anyway, and if they were to mess with the inputs to enter set a different langcode, as long as it's valid according to `langcodes`, it will just set the cookie but they won't see the language change. So there's no functional difference, it's just which one feels like the right source of truth.

Reply:
I had a thought that it could be worth trying to fall back to the two letter langcode here if the locales haven't been updated, but after some additional fiddling with this, I don't think it is. Django doesn't seem to load translations for languages that aren't defined in `settings.LANGUAGES` even if they're present in config and file structure, which means that you can set that language cookie to whatever language you want, but if it's not in `settings.LANGUAGES`, django won't load in the translations for it.

This also helps with my source-of-truth dilemma -- going to use `LANGUAGES` because that's what django seems to be using.